src/uproot/behaviors/TBranch.py

```diff
@@ -689,6 +689,211 @@ def foreach(start):
+    * str/bytes: same with an object-within-ROOT path, separated by a colon.
+      Example: ``"rel/file.root:tdirectory/ttree"``
+    * pathlib.Path: always interpreted as a filesystem path or URL only (no
+      object-within-ROOT path), regardless of whether there are any colons.
+      Examples: ``Path("rel:/file.root")``, ``Path("/abs/path:stuff.root")``
+    * glob syntax in str/bytes and pathlib.Path.
+      Examples: ``Path("rel/*.root")``, ``"/abs/*.root:tdirectory/ttree"``
+    * dict: keys are filesystem paths, values are objects-within-ROOT paths.
+      Example: ``{{"/data_v1/*.root": "ttree_v1", "/data_v2/*.root": "ttree_v2"}}``
+    * already-open TTree objects.
+    * iterables of the above.
+
+    Options (type; default):
+
+    * file_handler (:doc:`uproot.source.chunk.Source` class; :doc:`uproot.source.file.MemmapSource`)
+    * xrootd_handler (:doc:`uproot.source.chunk.Source` class; :doc:`uproot.source.xrootd.XRootDSource`)
+    * http_handler (:doc:`uproot.source.chunk.Source` class; :doc:`uproot.source.http.HTTPSource`)
+    * object_handler (:doc:`uproot.source.chunk.Source` class; :doc:`uproot.source.object.ObjectSource`)
+    * timeout (float for HTTP, int for XRootD; 30)
+    * max_num_elements (None or int; None)
+    * num_workers (int; 1)
+    * num_fallback_workers (int; 10)
+    * begin_chunk_size (memory_size; 512)
+    * minimal_ttree_metadata (bool; True)
+
+    Other file entry points:
+
+    * :doc:`uproot.reading.open`: opens one file to read any of its objects.
+    * :doc:`uproot.behaviors.TBranch.iterate`: iterates through chunks of
+      contiguous entries in ``TTrees``.
+    * :doc:`uproot.behaviors.TBranch.concatenate`: returns a single
+      concatenated array from ``TTrees``.
+
+    """
+    try:
+        import dask
+        import dask.array as da
+    except ModuleNotFoundError:
+        raise ModuleNotFoundError()
+
+    files = _regularize_files(files)
+    library = uproot.interpretation.library._regularize_library(library)
+    if library.name != "np":
+        raise NotImplementedError()
+
+    real_options = dict(options)
+    if "num_workers" not in real_options:
+        real_options["num_workers"] = 1
+    if "num_fallback_workers" not in real_options:
+        real_options["num_fallback_workers"] = 1
+
+    filter_branch = uproot._util.regularize_filter(filter_branch)
+
+    all_arrays = []
+    file_all_keys = []
+    n_array = []
+    count = 0
+
+    for file_path, object_path in files:
+        hasbranches = _regularize_object_path(
+            file_path, object_path, custom_classes, allow_missing, real_options
+        )
+        if hasbranches is not None:
+            count += 1
+            arrays = hasbranches.arrays(
+                expressions=expressions,
+                filter_name=filter_name,
+                filter_typename=filter_typename,
+                filter_branch=filter_branch,
+                library=library,
+            )
```

Comment:
Note: the [hasbranches.arrays](https://uproot.readthedocs.io/en/latest/uproot.behaviors.TBranch.HasBranches.html#arrays) function reads in all of the arrays, and since it's in a loop over all `file_path`s, it reads all arrays from all files. I see that this is only used below in line 887, to get the `dtype` of these arrays, but the `dtype`s should come from some source that does not require reading: [TBranch.interpretation](https://uproot.readthedocs.io/en/latest/uproot.behaviors.TBranch.TBranch.html#interpretation) (which, for pure NumPy arrays, is [AsDtype.to_dtype](https://uproot.readthedocs.io/en/latest/uproot.interpretation.numerical.AsDtype.html#to-dtype)).